src/Chainweb/Pact/PactService.hs

```diff
@@ -648,51 +672,51 @@ rewindTo mb act = do
 
     failNonGenesisOnEmptyDb = fail "impossible: playing non-genesis block to empty DB"
 
-    playFork cp bhdb payloadDb newH parentHash (lastBlockHeight, lastHash) =
-      flip catches exHandlers $ do
-          parentHeader <- liftIO $ lookupM bhdb parentHash
-          lastHeader <- liftIO $ lookupM bhdb lastHash
-          (!_, _, newBlocks) <-
-              liftIO $ collectForkBlocks bhdb lastHeader parentHeader
-          -- play fork blocks
-          V.mapM_ (fastForward payloadDb) newBlocks
-          -- play new block
-          restoreCheckpointer (Just (newH, parentHash)) >>= act
+    playFork cp bhdb payloadDb newH parentHash (lastBlockHeight, lastHash) = do
+        parentHeader <- liftIO $ lookupM bhdb parentHash
+        lastHeader <- findValidParent lastBlockHeight lastHash
+
+        (!_, _, newBlocks) <-
+            liftIO $ collectForkBlocks bhdb lastHeader parentHeader
+        -- play fork blocks
+        V.mapM_ (fastForward payloadDb) newBlocks
+        -- play new block
+        restoreCheckpointer (Just (newH, parentHash))
       where
-        exHandlers = [Handler handleTreeDbFailure, Handler handlePayloadFailure]
-        handleEx :: forall e . Exception e => e -> PactServiceM cas a
-        handleEx e =
-              (liftIO $ getBlockParent cp (lastBlockHeight, lastHash)) >>= \case
-                Nothing -> throwM e
-                Just hash -> playFork cp bhdb payloadDb newH parentHash
```

Comment:
I usually try to avoid recursion (or any larger long-running operations) in exception handers, because, I think, those are executed under a mask. Also when the code in the exception handler itself throws and has non-trivial exception handling, it seems that the overall behavior can become quite complex to understand.

But maybe I am just to paranoid :-)